RFC-0023-Unified-Memory-for-Pytorch.md

```diff
@@ -0,0 +1,127 @@
+     - This is true for all common types of copies, regardless of their direction, DtoH, HtoD, DtoD 
+   - In aten/src/ATen/native/cuda/Copy.cu::copy_kernel_cuda:
+     - No explicit cudaMemcpy in either direction 
+     - Set cudaMemPrefetchAsync() 
+   - Device to Host synchronization
+     - There is no concept of a GPU “stream” on the host so Device-to-Host copies have a required synchronization step attached. With UVM enabled, no explicit copy will be scheduled, but synchronization will still be required. 
+   - Device to Device
+     - If the device is the only tensor attribute being changed, the memadvise will be changed to reflect the new GPU. 
+
+ - .to(dtype, layout) (Or no UVM move on copy enabled) 
+
+   - The `_to_copy` function will operate as normal.  We will not interfere with backloading the storage of the at::empty() with existing pointers, and let a new copy of the tensor exist, and be returned. There is reasonable expectation that a user will want to cast a tensor for a single operation, and not lose the original data. 
+
+   - In `aten/src/ATen/native/cuda/Copy.cu::copy_kernel_cuda`:
+     - The tensor copy will occur as normal 
+     - set memadvise in the direction of the transfer 
+     - Set cudaMemPrefetchAsync() 
+
+   - Device to Host synchronization 
+     - There is no concept of a GPU “stream” on the host so Device-to-Host copies have a required synchronization step attached. With UVM enabled, a copy will occur with the destination tensor having the desired attribute changes requested. 
+   - Device to Device 
+     - If a device and additional attribute are being changed at the same time, a copy of the tensor will be made, and the necessary conversions will be applied as already designed. 
+
+
+## Additional Notes and Design Considerations
+Tuning Managed Memory with MemAdvise hints 
+
+ - Device allocations will use the managed memory APIs with initial memadvise hints for device mapping.  
+   - CudaMemAdviseSetPreferredLocation = current GPU device 
+   - CudaMemAdviseSetAccessedBy = unset 
+ - CPU allocations will be directed to the CachingManagedAllocator, but with initial memadvise hints for host mapping. 
+   - CudaMemAdviseSetPreferredLocation = cudaCpuDeviceId 
+   - CudaMemAdviseSetAccessedBy = current GPU device 
+
+### Optimal Block Sizes
+It is possible that there are different optimal block sizes when using managed memory due to default DMA sizes when migrating memory. Byte size boundaries for memory block division function to be determined based on prototyping 
+
+ - Currently, CUDACachingAllocator uses small and large block sizes: 
+   - Small: 2M 
+   - Large: 20M
+
+## FAQ 
+
+#### Should we have set_enabled_uvm(bool)? Will there be a use case where the user can set this to false? 
```

Comment:
This was the result of our "big switch" approach of enabling UVM mode. Since memory must be allocated using a different API, having it always on, or off made the design simpler.

Perhaps being able to mark a tensor as "managed" would help your use-case above. We could add a new flag (`managed=true`) to the to() function that would be able to cast a tensor to a managed tensor. We could still have the big switches (and a big switch disable), but those would just dictate the default value of the managed parameter here.

This would allow more user control on which tensors are managed. So:

recast (reallocate) a non-managed tensor to a managed tensor
```
tensor.to(device="cuda:0", managed=true)
```
The reverse would also work.